File: solutions/2716-prime-subtraction-operation/prime-subtraction-operation.java
class Solution {
    
    ArrayList<Integer> primes = new ArrayList<Integer>();
    
    public void getPrimes(){
        
        boolean[] visited = new boolean[10001];
        
        for(int i = 2; i <= 10000; i++){
            if(!visited[i]){
                visited[i] = true;
                primes.add(i);
                
                int j = i;
                while(j <= 10000){
                    visited[j] = true;
                    j += i;
                }
            }
        }
        
    }
    
    public boolean primeSubOperation(int[] nums) {
        getPrimes();
        
        for(int i = 0; i < nums.length; i++){
            int prime = 0;
            if(i == 0){
                for(int j : primes){
                    if(j >= nums[i]){
                        break;
                    }

                    prime = j;
                }

                nums[i] -= prime;

                // if(prime != 0)
                //     primes.remove((Integer)prime);
            }else{
                for(int j : primes){
                    if(j >= nums[i]){
                        break;
                    }

                    if(nums[i-1] < (nums[i] - j)){
                        prime = j;
                    }
                }

                nums[i] -= prime;

                // if(prime != 0)
                //     primes.remove((Integer)prime);
            }
        }

        System.out.println(Arrays.toString(nums));

        for(int i = 1; i < nums.length; i++){
            if(nums[i-1] >= nums[i]){
                return false;
            }
        }

        return true;
    }
}

File: solutions/2716-prime-subtraction-operation/prime-subtraction-operation.swift
class Solution {
    func primeSubOperation(_ nums: [Int]) -> Bool {
        var numbers = nums
        
        // Generate primes up to 1000
        var sieve = Array(repeating: true, count: 1001)
        sieve[0] = false
        sieve[1] = false
        
        for i in 2...Int(sqrt(Double(1000))) {
            if sieve[i] {
                var j = i * i
                while j <= 1000 {
                    sieve[j] = false
                    j += i
                }
            }
        }
        
        // Process from left to right
        for i in 0..<numbers.count {
            let target = i == 0 ? 0 : numbers[i-1]
            if numbers[i] <= target {
                return false
            }
            
            // Find largest prime to subtract while keeping array strictly increasing
            for p in (1...numbers[i]).reversed() where sieve[p] {
                let newVal = numbers[i] - p
                if newVal > target {
                    numbers[i] = newVal
                    break
                }
            }
        }
        
        return true
    }
}